spring-integration-jdbc/src/main/java/org/springframework/integration/jdbc/lock/JdbcLockRegistry.java

```diff
@@ -0,0 +1,215 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.integration.jdbc.lock;
+
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.Map;
+import java.util.Map.Entry;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.locks.Condition;
+import java.util.concurrent.locks.Lock;
+import java.util.concurrent.locks.ReentrantLock;
+
+import org.springframework.integration.support.locks.DefaultLockRegistry;
+import org.springframework.integration.support.locks.ExpirableLockRegistry;
+import org.springframework.integration.support.locks.LockRegistry;
+import org.springframework.integration.util.UUIDConverter;
+import org.springframework.util.Assert;
+
+/**
+ *
+ * A {@link LockRegistry} using a shared database to co-ordinate the locks. Provides the
+ * same semantics as the {@link DefaultLockRegistry}, but the locks taken will be global,
+ * as long as the underlying database supports the "serializable" isolation level in its
+ * transactions.
+ *
+ * @author Dave Syer
+ *
+ */
+public class JdbcLockRegistry implements ExpirableLockRegistry {
+
+	private final Map<String, JdbcLock> locks = new HashMap<String, JdbcLock>();
+
+	private JdbcClient client;
+
+	public JdbcLockRegistry(JdbcClient client) {
+		this.client = client;
+	}
+
+	@Override
+	public Lock obtain(Object lockKey) {
+		Assert.isInstanceOf(String.class, lockKey);
+		String path = pathFor((String) lockKey);
+		JdbcLock lock = this.locks.get(path);
+		if (lock == null) {
+			synchronized (this.locks) {
+				lock = this.locks.get(path);
+				if (lock == null) {
+					lock = new JdbcLock(this.client, path);
+					this.locks.put(path, lock);
+				}
+			}
+		}
+		return lock;
+	}
+
+	private String pathFor(String input) {
+		return input == null ? null : UUIDConverter.getUUID(input).toString();
+	}
+
+	@Override
+	public void expireUnusedOlderThan(long age) {
+		synchronized (this.locks) {
+			Iterator<Entry<String, JdbcLock>> iterator = this.locks.entrySet().iterator();
+			long now = System.currentTimeMillis();
+			while (iterator.hasNext()) {
+				Entry<String, JdbcLock> entry = iterator.next();
+				JdbcLock lock = entry.getValue();
+				if (now - lock.getLastUsed() > age && !lock.isAcquiredInThisProcess()) {
+					iterator.remove();
+				}
+			}
+		}
+	}
+
+	private static final class JdbcLock implements Lock {
+
+		private final JdbcClient mutex;
+
+		private final String path;
+
+		private volatile long lastUsed = System.currentTimeMillis();
+
+		private ReentrantLock delegate = new ReentrantLock();
+
+		private JdbcLock(JdbcClient client, String path) {
+			this.mutex = client;
+			this.path = path;
+		}
+
+		public long getLastUsed() {
+			return this.lastUsed;
+		}
+
+		@Override
+		public void lock() {
+			this.delegate.lock();
+			try {
+				boolean acquired = this.mutex.acquire(this.path);
```

Comment:
Shouldn't we catch here that `TransactionTimedOutException` (`@Transactional(isolation = Isolation.SERIALIZABLE, timeout = 1)`) and do `while(true)` to meet `Lock.lock()` contract?
```
* <p>If the lock is not available then the current thread becomes
* disabled for thread scheduling purposes and lies dormant until the
* lock has been acquired.
```

Don't tell me `ZookeeperLockRegistry` again. That might be not ideal example or just rely on the target `mutex` implementation. :smile: